# Nivell 1

## Pregunta 1

Executa el següent codi en una cel·la:

```java
byte b1 = (byte)127;
byte b2 = (byte)149;
println(b1);
println(b2);
```

Quins resultats obtens? **Per què**?

In [1]:
//Escriu i executa el codi aquí
byte b1 = (byte)127;
byte b2 = (byte)149;
println(b1);
println(b2);

127
-107


Resposta:

Byte es una tipus de dade que pot almacenar 1 byte de tamañ, osigui sols pot arribar a representar el numero +127, aixo te una consecuencia que si añadim un numero mes gran que +127 causi alteracions en el numero, en aquet cas sabem que el ultim bit es 1 ja que ens apareix -107 y es negatiu. 

## Pregunta 2

Executa el següent codi en una cel·la:

```java
var v1 = 10;
var v2 = 10.0;
println(v1);
println(v2);
```

De quin tipus de dada són les variables v1 i v2? **Per què**?

In [2]:
var v1 = 10;
var v2 = 10.0;
println(v1);
println(v2);

10
10.0


Resposta:

v1 es una variable de tipus integer, mestres la v2 es una variable de tipus double.
Tot aixo hu causa per el .0 que li indica que es un numero decimal i que el guardi amb decimals.
var es un tipus de variable indefinida, que hu fagui el programa sols quin tipus de dade encaixa amb el valor establit.

Ara, executa:

```java
var v3 = v1 + v2;
```
De quin tipus de dada és la variable v3? **Per què** (què està passant en fer-se la suma)?

In [5]:
var v3 = v1 + v2;
print(v3)

20.0

Resposta:
v3 es un tipus de variable de double, ja que el realitza una operacio es queda amb el tipus de dade que pugui complir els dos valors.

## Pregunta 3

Executa el següent codi en una cel·la:

```java
var b = 65;
println(b);
b = 'B';
println(b);

var a = 'A';
println(a);
a = 66;
println(a);
```

De quin tipus de dada són les variables b i a? **Per què**?

In [6]:
var b = 65;
println(b);
b = 'B';
println(b);

var a = 'A';
println(a);
a = 66;
println(a);

65
66
A
B


Resposta:

De tipus char (Caracter) y al colocar-li un numero al tipus de dades char hu relaciona per defecta al codi ASCII axio causa que cuan coloquem el numero 65 ens guardi la valor en char i sera la A.

## Exercici 1

#### Divisibilitat

Un número enter A **és divisible** per un altre número enter B quan el resultat de la seva divisió dóna també un número enter (és a dir, un número sense decimals, o amb decimals igual a 0). Per exemple:

In [7]:
//10 és divisible per 2
println(10.0 / 2);

//15 NO és divisible per 2
println(15.0 / 2);

//15 sí és divisible per 3
println(15.0 / 3);

5.0
7.5
5.0


Però la manera més ràpida de saber si un número enter A és divisible per un altre número enter B, no és fent-ne la divisió, sinó **calculant-ne el residu** de la divisió entera:

- Si el residu dóna 0, A és divisible per B
- Si el residu NO dóna 0, A NO és divisible per B

En programació, es calcula el residu amb l'operació anomenada **mòdul**, que s'expressa amb el símbol ```%```. Per exemple:

In [8]:
//El residu de 10 dividit entre 2 és 0, per tant 10 és divisible per 2
println(10 % 2);

0


In [9]:
//El residu de 15 dividit entre 2 NO és 0, per tant 15 NO és divisible per 2
println(15 % 2);

1


> **Tots els números enters són divisibles per sí mateixos i per 1**. Sigui N un número enter, sempre es compleix:

```
N % N == 0
N % 1 == 0
```

#### Números primers

Un número enter N és primer si només és divisible per sí mateix i per 1, és a dir, es compleix:

```
N % N == 0
N % 1 == 0
N % k != 0
```
**Per tot valor de k entre 2 i N - 1**. Si fos divisible per algun valor de k, llavors N no és primer.

#### Exercici

Escriu una funció, la signatura de la qual serà

```java
void isPrime(int n);
```
que imprimeixi ```true``` si n és un número primer positiu, o ```false``` si n és un número no primer o bé és 0, o bé és un enter negatiu.

Per exemple:

```java
isPrime(7); //Imprimirà true
isPrime(20); //Imprimirà false
isPrime(0); //Imprimirà false
isPrime(-7); //Imprimirà false
```

In [25]:
//Escriu i prova el teu codi aquí

void isPrime(int n){
    if (n % 2 != 0 && n != 0 && n > 0 ){
        println(true);
        }
    
        else{
        println(false);
    }
}
//Proves aquí
isPrime(7); 
isPrime(20); 
isPrime(0); 
isPrime(-7); 

true
false
false
false


## Exercici 2

Escriu una funció, la signatura de la qual serà

```java
int getMax(int[] n);
```

que retorni un enter igual al màxim de tots els enters que composen l'array.

In [55]:
//Escriu i prova el teu codi aquí

int getMax(int[] n){
    int suma= 0;
       for(int index = 0; index < n.length; index++){
           suma++;
    }
    return suma;
}
//Crea les teves proves aquí
int[] n= {1,3,2,4,2,8,6,8,2};

In [56]:
printf("La suma és igual a %d", getMax(n));

La suma és igual a 9

## Exercici 3

Escriu una funció, la signatura de la qual serà

```java
double getAverage(double[] d);
```

que retorni un double igual a la mitjana aritmètica de tots els elements que composen l'array.

In [101]:
//Escriu i prova el teu codi aquí

double getAverage(double[] d){
     double suma = 0.0;
     double suma2 = 0.0;
    for(int i = 0; i < d.length; i++){
        suma += d[i];
        suma2 = suma / d.length;
    }
    return suma2;
}

double[] d = {1.0,3.0,2.0,4.0,2.0,8.0,6.0,8.0};

In [102]:
printf("La suma és igual a %f", getAverage(d));

La suma és igual a 4,250000

In [98]:
print(suma2);

CompilationException: 